### PR TITLE
[shopsys] BreadcrumbGenerators are now automatically registered in BreadcrumbResolver

### LINKED ISSUE #919: There is no documentation about breadcrumb generation
### What is happening

I wasn't able to create new page on FE, because i didn't know how to work with breadcrumb. In documentation there isn't any mention how to work with that. 

### Expected result

In documentation should exist instructions how to work with breadcrumb.

### LINKED ISSUE #1090: Registering of BreadcrumbGenerators should be easier
### What is happening
If you want to register new BreadcrumbGenerators, you need to override BreadcrumbResolverFactory.

### Expected result
There should be some tag that would say, that this service is BreadcrumbGenerator and then it should be automatically registered. Same as it is in case of shopsys.friendly_url_provider.

Better way would be to tag services automatically by their classes interfaces:
- all classes implementing FriendlyUrlDataProviderInterface should be automatically tagged by shopsys.friendly_url_provider
- all classes implementing BreadcrumbGeneratorInterface should be automatically tagged by shopsys.breadcrumb_generator.

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Registering of new `BreadcrumbGenerator` was too painful. Users had to overwrite `BreadcrumbResolverFactory`, update `services.yml` etc.
|New feature| Yes 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| Yes 
|Fixes issues| closes #919 closes #1090 closes #1111 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

**Comment by PetrHeinz** [comment]
might be a duplicate of #1111

**Comment by TomasLudvik** [comment]
Hello @LukasHeinz, the #1111 is WIP but this one is complete development prepared for code review. That is why I prefer this one.

File: docs/introduction/friendly-url-breadcrumb-navigation.md [hunk 1]
```diff
@@ -31,3 +31,18 @@ The rest of the work is done automatically and URLs provided by these providers
         - if the name of the entity will be changed, the old URLs will now redirect to your new URLs
 - run `php phing friendly-urls-generate` to generate new URLs
 - visit some of provided URLs and check if everything works fine
+
+# Breadcrumb navigation
+
+All frontend routes include breadcrumb navigation in top of content page to ease navigation for your customers across your e-commerce platform.
```
Comment (vitek-rostislav) [inline docs/introduction/friendly-url-breadcrumb-navigation.md]:
```suggestion
All frontend routes include breadcrumb navigation in the top of the page to ease the navigation for your customers across your e-commerce site.
```

File: docs/introduction/friendly-url-breadcrumb-navigation.md [hunk 2]
```diff
@@ -31,3 +31,18 @@ The rest of the work is done automatically and URLs provided by these providers
...
+You might want to implement new `BreadcrumbGenerator` for your new frontend routes to tell application what navigation you want to display.
```
Comment (vitek-rostislav) [inline docs/introduction/friendly-url-breadcrumb-navigation.md]:
```suggestion
When adding a new page on the frontend, you need to implement new `BreadcrumbGenerator` for the new routes to tell the application how the navigation should be displayed.
```

File: docs/introduction/friendly-url-breadcrumb-navigation.md [hunk 3]
```diff
@@ -31,3 +31,18 @@ The rest of the work is done automatically and URLs provided by these providers
...
+
+## How to create new `BreadcrumbGenerator`
+
+- create new class with name ending with `BreadcrumbGenerator`
+- this class has to implement `BreadcrumbGeneratorInterface`
+- this interface requires you to implement two methods *(see `ArticleBreadcrumbGenerator` class as an example of the implementation)*:
+    - `getBreadcrumbItems` method that generates `BreadcrumbItems`
+        - these include displayed name, route and route parameters
+    - `getRouteNames` method where you have to provide names of routes where you want this navigation to appear
```
Comment (vitek-rostislav) [inline docs/introduction/friendly-url-breadcrumb-navigation.md]:
```suggestion
    - `getRouteNames` method where you have to provide names of the routes that you want to include in the navigation
```

File: docs/introduction/friendly-url-breadcrumb-navigation.md [hunk 4]
```diff
@@ -31,3 +31,18 @@ The rest of the work is done automatically and URLs provided by these providers
...
+All frontend routes include breadcrumb navigation in the top of the page to ease the navigation for your customers across your e-commerce site.
+When adding a new page on the frontend, you need to implement new `BreadcrumbGenerator` for the new routes to tell the application how the navigation should be displayed.
+
+## How to create new `BreadcrumbGenerator`
+
+- create new class with name ending with `BreadcrumbGenerator`
+- this class has to implement `BreadcrumbGeneratorInterface`
+- this interface requires you to implement two methods *(see `ArticleBreadcrumbGenerator` class as an example of the implementation)*:
+    - `getBreadcrumbItems` method that generates `BreadcrumbItems`
+        - these include displayed name, route and route parameters
```
Comment (vitek-rostislav) [inline docs/introduction/friendly-url-breadcrumb-navigation.md]:
```suggestion
        - these include displayed name, and may include route and route parameters if you want to make a link from the item
```

File: docs/introduction/friendly-url-breadcrumb-navigation.md [hunk 5]
```diff
@@ -31,3 +31,18 @@ The rest of the work is done automatically and URLs provided by these providers
...
+    - `getRouteNames` method where you have to provide names of the routes that you want to include in the navigation
```
Comment (vitek-rostislav) [inline docs/introduction/friendly-url-breadcrumb-navigation.md]:
```suggestion
    - `getRouteNames` method where you have to provide names of the routes for which you want to use the your breadcrumb generator
```

**Pull request closed without merging**